File: cpp/hot100/55.全排列.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        int n = nums.size();
        vector<int> path(n), on_path(n);
        auto dfs = [&](this auto && dfs, int i){
            if (i == n){
                ans.push_back(path);
                return ;
            }
            for (int j = 0; j < n; j++){
                if(!on_path[j]){
                    path[i] = nums[j];
                    on_path[j] = true;
                    dfs(i + 1);
                    on_path[j] = false;
                }
            }
        };
        dfs(0);
        return ans;
    }
};

File: python/hot100/55.全排列.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []
        n = len(nums)
        path = [0] * n
        def dfs(i, s):
            if i == n:
                ans.append(path.copy())
                return 
            for x in s:
                path[i] = x
                dfs(i + 1, s - {x})
        dfs(0, set(nums))
        return ans